crypto/x509/x509_vfy.c

```diff
@@ -1752,27 +1754,48 @@ static int internal_verify(X509_STORE_CTX *ctx)
      * is allowed to reset errors (at its own peril).
      */
     while (n >= 0) {
-        EVP_PKEY *pkey;
+        /*
+         * For each iteration of this loop:
+         * n is the subject depth
+         * xs is the subject cert, for which the signature is to be checked
+         * xi is the supposed issuer cert containing the public key to use
+         * Initially xs == xi if the last cert in the chain is self-issued.
+         */
+        /*
+         * If the issuer's public key is not available or its key usage does
+         * not support issuing the subject cert, report the issuer certificate
+         * and its depth (rather than n, the depth of the subject).
+         */
+        int issuer_depth = n + (xi == xs ? 0 : 1);
+
+        if (xs != xi) {
+            /* check key usage, but not for self-issued (typically root) cert */
+            int ret = x509_signing_allowed(xi, xs);
 
+            if (ret != X509_V_OK && !verify_cb_cert(ctx, xi, issuer_depth, ret))
+                return 0;
+        }
         /*
-         * Skip signature check for self signed certificates unless explicitly
-         * asked for.  It doesn't add any security and just wastes time.  If
-         * the issuer's public key is unusable, report the issuer certificate
-         * and its depth (rather than the depth of the subject).
+         * Skip signature check for self-issued certificates unless explicitly
+         * asked for because it does not add any security and just wastes time.
+         * We must not verify the signature if key usage prohibits signing.
```

Comment:
See my above quote form the RFC 5280, which so far had been neglected:
```
If the keyUsage extension is present, then the subject public key
   MUST NOT be used to verify signatures on certificates or CRLs unless
   the corresponding keyCertSign or cRLSign bit is set
```

Reply:
It this clear now, together with the further extended code comments and my new comment in the above review discussion?

Reply:
This is my summary of the proposed updated commit:

If a self-issued root/EE cert is last in chain we verify its signature only if `X509_V_FLAG_CHECK_SS_SIGNATURE` is requested.
(This has been the semantics so far.)

What this commit changes is:
Even in this case skip the signature verification if the cert key usage prohibits signing.
Make clear that we must check key usage when verifying the signature of
the next cert and we must not verify a signature if key usage forbids it.
Add some comments for making internal_verify() easier to understand.
Update the documentation of `X509_V_FLAG_CHECK_SS_SIGNATURE` accordingly.

Reply:
> What this commit changes is:
> Even in this case skip the signature verification if the cert key usage prohibits signing.

I am not sure that's what users of `X509_V_FLAG_CHECK_SS_SIGNATURE` would expect.  For a self-signed EE certificate, that does not have `kuCertSign` (intended to be a leaf only, not a CA), a self signature check is still possible, and would likely be expected if `X509_V_FLAG_CHECK_SS_SIGNATURE` is set.  The EE self-signed case is entirely outside the scope of RFC5280.  It can only be trusted by being directly bound as fit for a particular service without any CA involvement.  So we get to set the verification rules subject to reasonable user expectations and our own documentation.  So I'd like to ask for this to be considered more carefully, and a brief rationale written down one way or the other...

> Make clear that we must check key usage when verifying the signature of
> the next cert and we must not verify a signature if key usage forbids it.
> Add some comments for making internal_verify() easier to understand.
> Update the documentation of `X509_V_FLAG_CHECK_SS_SIGNATURE` accordingly.

Reply:
> I am not sure that's what users of X509_V_FLAG_CHECK_SS_SIGNATURE would expect. For a self-signed EE certificate, that does not have kuCertSign (intended to be a leaf only, not a CA), a self signature check is still possible, 

The self-signature check is technically possible in most cases, but not if the public key simply does not support signing,
and this is one reason why the key usage keyCertSign is prohibited.
An example is the cert in https://tools.ietf.org/html/draft-ietf-curdle-pkix-04#section-10.2
where the private key corresponding to the included X25519 public key simply cannot be used for signing,
so trying to verify the cert signature makes no sense. 

> and would likely be expected if X509_V_FLAG_CHECK_SS_SIGNATURE is set.

Yes, but there are certs where this expectation cannot be met
or at least according to the given key usage restrictions the verification is prohibited.

Reply:
> The EE self-signed case is entirely outside the scope of RFC5280. 

This would surprise me - can you provide a proof for this statement?
I cannot find any such scope restriction on/within RFC 5280.

Reply:
5280 has no concept of self-signed EE certificates.  It only discusses self-signed CA certificates:
```
   This specification covers two classes of certificates: CA
   certificates and end entity certificates.  CA certificates may be
   further divided into three classes: cross-certificates, self-issued
   certificates, and self-signed certificates.  Cross-certificates are
   CA certificates in which the issuer and subject are different
   entities.  Cross-certificates describe a trust relationship between
   the two CAs.  Self-issued certificates are CA certificates in which
   the issuer and subject are the same entity.  Self-issued certificates
   are generated to support changes in policy or operations.  Self-
   signed certificates are self-issued certificates where the digital
   signature may be verified by the public key bound into the
   certificate.  Self-signed certificates are used to convey a public
   key for use to begin certification paths.  End entity certificates
   are issued to subjects that are not authorized to issue certificates.
```
And yet, many users of OpenSSL deploy self-signed EE certificates
that are in no way intended to be used as CAs that issue certificates
for other subject names.  Furthermore, in 5280 a self-signed CA
certificate is just a public key container.  There is no process in
5280 describing verification of such certificates.
```
   When the trust anchor is provided in the form of a self-signed
   certificate, this self-signed certificate is not included as part of
   the prospective certification path.  Information about trust anchors
   is provided as inputs to the certification path validation algorithm
   (Section 6.1.1).
```

So my question is what to do about EE self-signed certs.  Should the self-signature
be checked when the `X509_V_FLAG_CHECK_SS_SIGNATURE` flag is set?  I don't
claim to have the definitive answer, but I do claim that no answer is or can be given
in RFC5280.  If I'm right, then we need to decide what the right answer (least-surprise)
answer is for users of OpenSSL.  It may well be to ignore keyUsage when verifying EE
self-signatures, and check the signature anyway, when `X509_V_FLAG_CHECK_SS_SIGNATURE`
is set.

Reply:
As I wrote above, the topic is not necessarily self-signed EE certs, but more generally self-issued EE certs.

And the definition of EE certs in 5280 does not refer to self-issued not self-signed. So it does note exclude either of those types of EE certs.

Reply:
Self-issued certs do not carry an explicit indication whether they are self-signed or not.
The only indication that one has is whether their key usage permits self-signing.

Reply:
In my view the `X509_V_FLAG_CHECK_SS_SIGNATURE` user flag should be interpreted as:

Verify the signature of the last cert in the chain as self-signature where this is appropriate.
This typically is the case of self-signed root CA certs, 
but also for some self-issued EE certs, namely for those where the key usage allows certificate signing.

As mentioned, there are self-issued EE certs where the type of the public key is algorithmically unsuitable for signing (for instance, pure key agreement keys such as DH), so attempting to still verify "self-signature" would fail deeply.